Add cluster smoke test job and run it during upgrades

codificat opened the pull request:
This adds a new job to run smoke tests on a cluster, and calls that job during the upgrade job to run a smoke test after a cluster has been upgraded.

The PR also contains some minor changes to the smoke test script itself - one of them is actually required: `oc get pod | grep Running` is prone to races resulting on a script failure, and instead of adding another `sleep` this uses `oc rollout status -w` instead which should be more reliable and appropriate.

Also taking the opportunity to update an unrelated job: just fixing the destination address for notifications of key rotation failures.

codificat commented on jobs/cluster/rotate-log-access-key/Jenkinsfile:
```diff
@@ -43,7 +43,7 @@ node('openshift-build-1') {
                     [
                         $class: 'hudson.model.StringParameterDefinition',
-                        defaultValue: 'pep@redhat.com',
+                        defaultValue: 'jupierce@redhat.com,pep@redhat.com',
```
This change is unrelated to the topic of the PR, but thought I'd hijack the PR a bit to include just this minor fix...

codificat commented on jobs/cluster/smoke-test/Jenkinsfile:
```diff
@@ -0,0 +1,86 @@
+                        $class: 'hudson.model.BooleanParameterDefinition',
+                        defaultValue: false,
+                        description: 'Send email notifications',
+                        name: 'MAIL_RESULTS'
```
So: this disables email notifications by default (i.e. no emails would be sent on success or failure). Should we have a default of `true` here? I set it to `false` because the nature of the job seems to me it might not be worth sending notifications for a single smoke test run...

Moreover: the upgrade job calls this one with email notifications explicitly disabled too, because it handles its own email notifications.

jupierce replied:
Seems like a reasonable way to start.

codificat commented on jobs/cluster/upgrade/Jenkinsfile:
```diff
@@ -85,7 +85,13 @@ node('openshift-build-1') {
             }
 
             stage( "smoketest" ) {
-                // run smoketest job here, try{}catch to determine if it failed and  include details of job in success email
```
I did not include another try/catch specifically for the smoke test here - there's a surrounding try/catch block already.

So we could just set `propagate: true` if we wanted to fail the upgrade job on smoke test failure. But instead we probably want to check the result and throw a custom error...

About details of the smoke test itself: I believe that a URL pointing to the specific smoke test job run is enough here.

jupierce replied:
The smoketest should be informational. The upgrade has already happened by this point, so we want to finish the pipeline. However, the "success" email we sent out should include a highly visible warning that smoketests have failed along with a link to the failing smoketest job's output. This is why the original smoketest was counting relatively minor infractions like nodes being unschedulable. 

So, if result != SUCCESS, I suggest just setting a flag that can be interpreted by mail_success. mail_success can then tailor the output to include the warning and the URL.

stevekuznetsov commented on tower-scripts/bin/aos-cd-cluster-smoke-test.sh:
```diff
@@ -63,24 +63,24 @@ fi
+    set -ex
+    PROJ_RANDOM=aos-cd-smoketest-$(shuf -i 100000-999999 -n 1)
+    oc new-project ${PROJ_RANDOM}
     sleep 10  # Allow project to initialize
```
This should try `oc get projects` until we see the project show up instead of a blind sleep

kargakis replied:
Agreed.

stevekuznetsov commented on tower-scripts/bin/aos-cd-cluster-smoke-test.sh:
```diff
@@ -29,7 +29,7 @@ if [ "$?" != "0" ]; then
 fi
 
 if oc get nodes -l type=infra | grep -e SchedulingDisabled -e NotReady; then
```
This is as fragile as `oc get pods | grep Running`

kargakis replied:
Why don't we make the node schedulable if it's unschedulable? Also, I would retry for some time on NotReady.

codificat replied:
I think that fixing stuff is not what the smoke test is for - if there's a problem, we want to highlight it, not try to fix it here as part of the test.

So, the question is if we should fail or not based on having *any* non-schedulable or NotReady node. Probably we don't... so we could just write down the warning; but then no one would read it probably (buried down in logs).

I'm thinking we should just remove this part of the smoke test (check for NotReady/unschedulable nodes). Note that the cluster status post-upgrade is already reported by the upgrade job, and this includes output of "oc get node".

kargakis commented on tower-scripts/bin/aos-cd-cluster-smoke-test.sh:
```diff
@@ -63,24 +63,24 @@ fi
     sleep 60  # Allow build to start
     timeout 10m oc logs -f bc/ruby-hello-world
-    oc get pods | grep "Running"
+    oc rollout status dc/ruby-hello-world -w
```
-w is true by default, and I don't think we will ever break it

codificat commented:
As mentioned I updated this so that the upgrade job doesn't fail due to smoke test failures but it adds a visible warning in the notification instead.

Also squashed and rebased (including the move from jobs/cluster to jobs/starter).

I still left out the check for non-schedulable/non-ready nodes because I believe that for large enough clusters is likely to have a tendency for false positives... also, the notification email, as part of cluster status info, already includes a list of all nodes and their statuses. Should I add that check back though?

jupierce commented on jobs/starter/smoke-test/Jenkinsfile:
```diff
@@ -0,0 +1,86 @@
+    )
+
+    def deploylib = load( "pipeline-scripts/deploylib.groovy" )
+    deploylib.initialize()
```
Cluster jobs have changed a bit while you have worked this card.

Populating cluster names:
https://github.com/openshift/aos-cd-jobs/blob/master/jobs/starter/perf/Jenkinsfile#L14-L17
https://github.com/openshift/aos-cd-jobs/blob/master/jobs/starter/perf/Jenkinsfile#L28

Initializing deploylib:
https://github.com/openshift/aos-cd-jobs/blob/master/jobs/starter/perf/Jenkinsfile#L39

Setting up ssh agent (CLUSTER_ENV instead of _NAME): 
https://github.com/openshift/aos-cd-jobs/blob/master/jobs/starter/perf/Jenkinsfile#L51

codificat replied:
Sorry I missed these, will update...

jupierce replied:
You didn't miss them, they changed underneath you :)

jupierce commented on jobs/starter/upgrade/Jenkinsfile:
```diff
@@ -1,16 +1,19 @@
+    body = "Cluster ${CLUSTER_NAME} upgrade details:\n"
+    if ( warn ) {
+        body += "\nWARNING: post-upgrade smoke test was not successful!\n"
+        body += "WARNING: smoke test job details: ${warn}\n"
```
Redundant "WARNING" ?

codificat replied:
This was intentional, for visual purposes... I thotught it makes it more obvious and easier to read/spot the smoke test URL vs the upgrade job URL which I also moved to the top for [hopefully] easier access.

A notification with smoke test warnings would look like this:

```
Cluster CLUSTER_NAME upgrade details:

WARNING: post-upgrade smoke test was not successful!
WARNING: smoke test job details: https://jenkins-jenkins.192.168.122.78.nip.io/job/smoketest/25/

Upgrade Jenkins job: https://jenkins-jenkins.192.168.122.78.nip.io/job/test/7/

<cluster status>

Smoke test results: FAILURE
```

I can leave the upgrade job URL at the bottom and supress the second WARNING line (leaving the smoke test URL only) if preferred

jupierce replied:
Maybe combining the lines is sufficient. I just don't want it to look as scary :)
WARNING: Post-upgrade smoke test was not successful: URL

jupierce commented:
@codificat Your point is well taken, but I don't consider a significant number of nodes being unschedulable/not-ready as being a false positive -- even in our large clusters. Experience has shown it to be an important signal for someone to check up on the health of the cluster. Perhaps we only fail the test if > 2 meet this condition. That gives us room for a little noise.

codificat commented:
New rebase/squash with the recent suggested changes:

- Keep the check for non-ready nodes (I didn't add a threshold)
- Make the smoke test failure warning less scary
- Update to match new cluster selection and library handling
